moveit_ros/moveit_servo/src/servo_calcs.cpp

```diff
@@ -188,17 +193,84 @@ void ServoCalcs::start()
   initial_joint_trajectory->points.push_back(point);
   last_sent_command_ = std::move(initial_joint_trajectory);
 
-  // Set up timer for calculation callback
-  timer_ = node_->create_wall_timer(std::chrono::duration<double>(period_), std::bind(&ServoCalcs::run, this));
-
   current_state_ = planning_scene_monitor_->getStateMonitor()->getCurrentState();
   tf_moveit_to_ee_frame_ = current_state_->getGlobalLinkTransform(parameters_->planning_frame).inverse() *
                            current_state_->getGlobalLinkTransform(parameters_->ee_frame_name);
   tf_moveit_to_robot_cmd_frame_ = current_state_->getGlobalLinkTransform(parameters_->planning_frame).inverse() *
                                   current_state_->getGlobalLinkTransform(parameters_->robot_link_command_frame);
+
+  stop_requested_ = false;
+  thread_ = std::thread([this] { mainCalcLoop(); });
+  new_input_cmd_ = false;
+}
+
+void ServoCalcs::stop()
+{
+  // Request stop
+  stop_requested_ = true;
+
+  // Notify condition variable in case the thread is blocked on it
+  {
+    // scope so the mutex is unlocked after so the thread can continue
+    // and therefore be joinable
+    const std::lock_guard<std::mutex> lock(input_mutex_);
+    new_input_cmd_ = false;
+    input_cv_.notify_all();
+  }
+
+  // Join the thread
+  if (thread_.joinable())
+  {
+    thread_.join();
+  }
 }
 
-void ServoCalcs::run()
+void ServoCalcs::mainCalcLoop()
+{
+  rclcpp::Rate rate(1.0 / parameters_->publish_period);
+
+  while (rclcpp::ok() && !stop_requested_)
+  {
+    // lock the input state mutex
+    std::unique_lock<std::mutex> input_lock(input_mutex_);
+
+    // low latency mode
+    if (parameters_->low_latency_mode)
+    {
+      input_cv_.wait(input_lock, [this] { return (new_input_cmd_ || stop_requested_); });
+
+      // break out of the loop if stop was requested
+      if (stop_requested_)
+        break;
+    }
```

Comment:
This chunk of code resulted in a weird bug that caused large jumps when restarting servo.  The fix is here: https://github.com/ros-planning/moveit/pull/2496/files